code/scripts/espurna_utils/bsec.py

```diff
@@ -0,0 +1,107 @@
+import os
+
+def bsec_inject_patcher(env):
+    platform = env.PioPlatform()
+    framework_dir = platform.get_package_dir("framework-arduinoespressif8266")
+    bsec_dir = os.path.join(
+        "${PROJECT_LIBDEPS_DIR}", "${PIOENV}", "BSEC Software Library", "src", "esp8266"
+    )
+
+    env.Append(LIBPATH=["${PROJECT_LIBDEPS_DIR}/${PIOENV}/BSEC Software Library/src/esp8266"])
+    env.Append(LIBS=["algobsec"])
+
+    patchflag_path = os.path.join(env.subst(bsec_dir), ".patching-done")
+
+    if os.path.isfile(patchflag_path):
+      print("Archive libalgosec.a has already been patched")
+      return
+
+    toolchain_ar = os.path.join(
+        platform.get_package_dir("toolchain-xtensa"), "bin", "xtensa-lx106-elf-ar"
+    )
+
+    # Backup libalgobsec.a into libalgobsec.a.orig.
+    backup_action = env.VerboseAction(
+        " ".join(
+            [
+               '/bin/mv',
+                os.path.join(bsec_dir.replace(' ', '\\ '), "libalgobsec.a"),
+                os.path.join(bsec_dir.replace(' ', '\\ '), "libalgobsec.a.orig")
+            ]
+        ),
+        "Backing up BSEC libalgosec.a to libalgosec.a.orig",
+    )
+
+    # Extract libalgobsec.a into *.o files.
+    extract_action = env.VerboseAction(
+        " ".join(
+            [
+               toolchain_ar,
```

Comment:
@mcspr I really dislike the fact that files are extract to current dir... it's a mess to maintain. Do you have any other idea for this? I don't see `ar` support for setting an output directory.

Reply:
Re `tempfile.mkdtemp()`, *I think* we are pretty safe to chdir in and out if we patch the source right away and don't try to create dependent actions.

*(I opened commit directly... sry)*

Reply:
Re2:

https://scons.org/doc/3.1.2/HTML/scons-man.html#f-Command
Is a 'Builder' variant that accepts action, but also allows chdir='...'. Not sure how that works, yet, because it implies target<->source dependency (so that Scons can track if file is built / up-to-date / etc., it will hash source and re-build when it changes)
Only option I had in mind was to just call a subprocess and chdir there